File: FsharpTodoApp.Infrastructure/Features/Todo/DataModels/TodoDataModel.fs
namespace FsharpTodoApp.Infrastructure.Features.Todo.DataModels

open FsharpTodoApp.Infrastructure.Persistence.DataModels
open FsharpTodoApp.Domain.Features.Todo.Entities
open FsharpTodoApp.Domain.Features.Todo.ValueObjects
open FsharpTodoApp.Persistence.DataModels

module TodoDataModel =
    let dehydrate (dataModel: TodoDataModel) (domain: TodoEntity) =
        domain.Base |> DataModelBase.dehydrate dataModel

        dataModel.Title <- domain.Title |> TodoTitle.value
        dataModel.Description <- domain.Description |> TodoDescription.value |> Option.defaultValue null
        dataModel.DueDate <- domain.DueDate |> TodoDueDate.value |> Option.toNullable
        dataModel.Status <- domain.Status |> TodoStatus.value |> TodoStatus.toEnum
        dataModel.Assignee <- domain.Assignee |> TodoAssignee.value |> Option.map _.UserName |> Option.defaultValue null
        dataModel.Reviewer <- domain.Reviewer |> TodoReviewer.value |> Option.map _.UserName |> Option.defaultValue null

File: FsharpTodoApp.Persistence/DataModels/TodoDataModel.cs
using FsharpTodoApp.Domain.Features.Todo.Enums;
using Microsoft.EntityFrameworkCore;

namespace FsharpTodoApp.Persistence.DataModels;

public class TodoDataModel : DataModelBase
{
    public string Title { get; set; } = string.Empty;
    public string? Description { get; set; }
    public DateTime? DueDate { get; set; }
    public TodoStatusEnum Status { get; set; }
    public string? Assignee { get; set; }
    public string? Reviewer { get; set; }

    public static void OnModelCreating(ModelBuilder modelBuilder)
    {
        OnModelCreating<TodoDataModel>(modelBuilder, "Todos");

        modelBuilder.Entity<TodoDataModel>()
            .Property(e => e.Status)
            .HasConversion(
                v => v.ToString(),
                v => Enum.Parse<TodoStatusEnum>(v));
    }
}
